src/language/providers/closeUtils.ts

```diff
@@ -0,0 +1,495 @@
+      '<' + nsPrefix + items[i],
+      nextOpenTagPos
+    )) > -1
+  ) {
+    openTagArray.push(nextOpenTagPos)
+
+    if ((nextCloseCharPos = triggerText.indexOf('>', nextOpenTagPos)) > -1) {
+      //if tag is self closing remove it from the openTagArray
+      if (triggerText.substring(nextCloseCharPos - 1, 2) === '/>') {
+        openTagArray.splice(-1, 1)
+      }
+
+      nextOpenTagPos = nextOpenTagPos + 1
+    }
+  }
+
+  while (
+    (nextCloseTagPos = triggerText.indexOf(
+      '</' + nsPrefix + items[i],
+      nextCloseTagPos
+    )) > -1
+  ) {
+    closeTagArray.push(nextCloseTagPos)
+    nextCloseTagPos = nextCloseTagPos + 1
+  }
+
+  if (openTagArray.length > closeTagArray.length) {
+    if (triggerPos > closeTagArray[closeTagArray.length]) {
+      return items[i]
+    }
+
+    while (closeTagArray.length > 0 && openTagArray.length > 0) {
+      //if the closing tag is before the trigger position
+      //and there is an opening tag before that closing tag
+      //remove the last opening tag before that first closing tag
+      //from the array otherwise remove the closing tag from
+      //the array
+      ic = 0
+      io = openTagArray.length
+
+      while (ic < closeTagArray.length) {
+        if (closeTagArray[ic] <= triggerPos) {
+          while (io < 0) {
+            if (openTagArray[io] < closeTagArray[ic]) {
+              closeTagArray.splice(ic, 1)
+              openTagArray.splice(io, 1)
+              io = openTagArray.length
+              break
+            }
+
+            --io
+          }
+        }
+
+        ++ic
+      }
+
+      //if the opening tag is after the trigger and the
+      //closing tag is after the opening tag remove them
+      //both from the array otherwise remove the opening tag
+      io = 0
+      ic = 0
+
+      while (io < openTagArray.length) {
+        if (openTagArray[0] >= triggerPos) {
+          while (ic < closeTagArray.length) {
+            if (closeTagArray[ic] > openTagArray[io]) {
+              closeTagArray.splice(io, 1)
+              openTagArray.splice(io, 1)
+              ic = 0
+              break
+            }
+
+            ++ic
+          }
+        }
+
+        ++io
+      }
+
+      //if the last closing tag is after the trigger position
+      //and the first opening tag is before the trigger position
+      //remove them both
+      ic = closeTagArray.length
+      io = 0
+
+      while (ic >= 0) {
+        if (closeTagArray[ic] >= triggerPos) {
+          while (io < openTagArray.length) {
+            if (openTagArray[io] < triggerPos) {
+              openTagArray.splice(io, 1)
+              closeTagArray.splice(ic, 1)
+              io = 0
+              break
+            }
+
+            ++io
+          }
+        }
+
+        --ic
+      }
+    }
+
+    //if there are any open tags left return the tag item
+    if (openTagArray.length > 0) {
+      return items[i]
+    }
+  }
+
+  return undefined
+}
+
+export function getItemsForLineLT2(
```

Comment:
@rt320 Same here, this should probably be split up more but it was hard for to follow some of the looping going so it would be best that possibly you handle this since you know it better